Mathlib/GroupTheory/Perm/MaximalSubgroups.lean

```diff
@@ -0,0 +1,430 @@
+import Mathlib.GroupTheory.SpecificGroups.Cyclic
+import Mathlib.GroupTheory.Subgroup.Simple
+import Mathlib.GroupTheory.GroupAction.SubMulAction.OfFixingSubgroup
+
+/-! # Maximal subgroups of the symmetric groups
+
+* `Equiv.Perm.isCoatom_stabilizer`: when `s : set α` is
+  not empty, nor its complementary subset,
+  and if the cardinality of `s` is not half of that of `α`,
+  then `MulAction.stabilizer (Equiv.Perm α) s` is
+  a maximal subgroup of the symmetric group `Equiv.Perm α`.
+
+  This is the *intransitive case* of the O'Nan-Scott classification.
+
+## TODO
+
+  * Appplication to primitivity of the action
+    of `Equiv.Perm α` on finite combinations of `α`.
+
+  * Formalize the other cases of the classification.
+  The next one should be the *imprimitive case*.
+
+## Reference
+
+The argument is taken from [M. Liebeck, C. Praeger, J. Saxl,
+*A classification of the maximal subgroups of the finite
+alternating and symmetric groups*, 1987][LiebeckPraegerSaxl-1987].
+-/
+
+open scoped Pointwise
+
+open Set
+
+theorem Subtype.image_preimage_of_val {α : Type*} {s B : Set α} (h : B ⊆ s) :
+    Subtype.val '' (Subtype.val ⁻¹' B : Set s) = B := by
+  rw [image_preimage_eq_range_inter, range_coe, inter_eq_self_of_subset_right h]
+
+namespace Equiv.Perm
+
+open MulAction
+
+variable (G : Type*) [Group G] {α : Type*} [MulAction G α]
+
+theorem IsPretransitive.of_partition {s : Set α}
+    (hs : ∀ a ∈ s, ∀ b ∈ s, ∃ g : G, g • a = b)
+    (hs' : ∀ a ∈ sᶜ, ∀ b ∈ sᶜ, ∃ g : G, g • a = b)
+    (hG : stabilizer G s ≠ ⊤) :
+    IsPretransitive G α := by
+  suffices ∃ (a b : α) (g : G), a ∈ s ∧ b ∈ sᶜ ∧ g • a = b by
+    obtain ⟨a, b, g, ha, hb, hgab⟩ := this
+    rw [isPretransitive_iff_base a]
+    intro x
+    by_cases hx : x ∈ s
+    · exact hs a ha x hx
+    · rw [← Set.mem_compl_iff] at hx
+      obtain ⟨k, hk⟩ := hs' b hb x hx
+      use k * g
+      rw [mul_smul, hgab, hk]
+  by_contra! hyp
+  apply hG
+  rw [eq_top_iff, le_stabilizer_iff_smul_le]
+  intro g _
+  rintro b ⟨a, ha, hgab⟩
+  by_contra hb
+  exact hyp a b g ha (Set.mem_compl hb) hgab
+
+theorem swap_mem_stabilizer [DecidableEq α]
+    {a b : α} {s : Set α} (ha : a ∈ s) (hb : b ∈ s) :
+    swap a b ∈ stabilizer (Perm α) s := by
+  suffices swap a b • s ⊆ s by
+    rw [mem_stabilizer_iff]
+    apply Set.Subset.antisymm this
+    exact Set.subset_smul_set_iff.mpr this
+  rintro _ ⟨x, hx, rfl⟩
+  simp only [Perm.smul_def]
+  by_cases hxa : x = a
+  · rwa [hxa, swap_apply_left]
+  by_cases hxb : x = b
+  · rwa [hxb, swap_apply_right]
+  · rwa [swap_apply_of_ne_of_ne hxa hxb]
+
+theorem moves_in
+    (G : Subgroup (Perm α)) (t : Set α) (hGt : stabilizer (Perm α) t < G) :
+    ∀ a ∈ t, ∀ b ∈ t, ∃ g : G, g • a = b := by
+  classical
+  intro a ha b hb
+  use ⟨swap a b, ?_⟩
+  · change swap a b • a = b
+    simp only [Perm.smul_def, swap_apply_left]
+  · apply le_of_lt hGt
+    apply swap_mem_stabilizer ha hb
+
+theorem stabilizer_ne_top (s : Set α) (hs : s.Nonempty) (hsc : sᶜ.Nonempty) :
```

Comment:
```suggestion
theorem stabilizer_ne_top_of_nonempty_of_nonempty_compl (s : Set α) (hs : s.Nonempty) (hsc : sᶜ.Nonempty) :
```
or something similar.  Otherwise, the name is too generic.